### PR TITLE
feature: Add config file support (CY-5214)

### PR DESCRIPTION
**Pull request by vascorsd**
Adding support for some spectral configuration files: .spectral{json, yaml, yml}. Others are excluded.

We are also only ensuring that known rules/patterns names are returned by the tool. If the need arises we can iterate on the specifics.

### CODE REVIEW

File: src/spectralRulesetLoader.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                console.debug(`paths: ${paths}`)
+
+                const _require = (id: string): unknown => require(require.resolve(id, {paths}))
+
+                Function('module, require', source)(m, _require)
+
+                return m.exports
+            })
+    } else {
+        console.debug(`could not process configuration file.`)
```
Comment (lventura-codacy) [inline src/spectralRulesetLoader.ts:37]: Should this be a warning?

Comment (vascorsd) [reply src/spectralRulesetLoader.ts]: Changed the way logging is done, so only done locally when passing DEBUG=1 to the tool runtime.

File: src/configExtractor.ts [hunk 2]
```diff
@@ -1,16 +1,16 @@
-import { Codacyrc } from "codacy-seed"
+import {Codacyrc} from "codacy-seed"
 import * as glob from "glob"
 
-import { toolName } from "./toolMetadata"
+import {toolName} from "./toolMetadata"
 
-export async function extractPatternIdsToApply(
+export async function extractPatternsFromRC(
```
Comment (lolgab) [inline src/configExtractor.ts:6]: It's hard to understand what `RC` means in my opinion.
Consider something like:
```suggestion
export async function extractPatternsFromCodacyrc(
```
which is also consistent with how we call the `codacyrc` file in other tools.

File: src/configExtractor.ts [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
     codacyrc?: Codacyrc
 ): Promise<string[] | undefined> {
     const tool = codacyrc?.tools ? codacyrc?.tools.find((codacyTool) => codacyTool.name === toolName) : undefined
     return tool?.patterns?.map(pattern => pattern.patternId)
 }
 
-export async function extractFiles(
+export async function extractFilesFromRC(
```
Comment (lolgab) [inline src/configExtractor.ts:13]: Same here.

File: src/engineImpl.ts [hunk 4]
```diff
@@ -1,69 +1,184 @@
-import { Document, Ruleset, RulesetDefinition,Spectral } from "@stoplight/spectral-core"
-import { FileRuleDefinition } from "@stoplight/spectral-core/dist/ruleset/types"
-import { Json,Yaml } from "@stoplight/spectral-parsers"
-import { asyncapi,oas } from "@stoplight/spectral-rulesets";
-import { Codacyrc, Engine, ToolResult } from "codacy-seed"
-import { readFile } from "codacy-seed"
+import {Document, Ruleset, RulesetDefinition, Spectral} from "@stoplight/spectral-core"
+import {FileRuleDefinition} from "@stoplight/spectral-core/dist/ruleset/types"
+import {Json, Yaml} from "@stoplight/spectral-parsers"
+import {asyncapi, oas} from "@stoplight/spectral-rulesets"
```
Comment (lolgab) [inline src/engineImpl.ts:4]: What's happening with the formatting? Did you change the formatter settings?

File: src/engineImpl.ts [hunk 5]
```diff
@@ -1,69 +1,184 @@
...
+            log("configuring spectral to use all recommended rules...")
+
+            s.setRuleset({
+                extends: [
+                    [asyncapi as RulesetDefinition, 'recommended'],
+                    [oas as RulesetDefinition, 'recommended'],
+                ]
+            })
+        } else {
+            log("configuring spectral to use explicit rules defined in codacyrc...")
```
Comment (lolgab) [inline src/engineImpl.ts:125]: What stream does `log` use? stderr?

File: src/logging.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export function log(msg: string) {
```
Comment (lolgab) [inline src/logging.ts:1]: Maybe `log` can be `debug`? Since it uses the `console.debug`?

**Review by lolgab — CHANGES_REQUESTED** [review]
Some little comments, otherwise LGTM!
Good job 👍
I really liked the commenting with great details!

File: src/logging.ts (hunk 6 above)
Comment (vascorsd) [reply src/logging.ts]: Changed to debug.

File: src/configExtractor.ts (hunk 3 above)
Comment (vascorsd) [reply src/configExtractor.ts]: Changed.

**Review by machadoit — APPROVED** [review]
LGTM!

**Review by lolgab — APPROVED** [review]

**Pull request merged**